Disable pref if AppIndicator Support not detected

cskeogh opened the pull request:
If running Linux and Gnome Desktop and AppIndicator Support [1] support is not detected, disable the preference `Leave app running in the notification area` and set to false. Style the disabled preference (Linux only). Style is suitable for normal mode and dark mode. Uses an environment variable as suggested by @nathanlesage [2]

[1] https://extensions.gnome.org/extension/615/appindicator-support/
[2] https://github.com/Zettlr/Zettlr/issues/175#issuecomment-814900960

Passes  `yarn lint`.

Closes #13

cskeogh committed:
Disable pref if AppIndicator Support not detected

If running Linux and Gnome Desktop and AppIndicator Support [1] support
is not detected, disable the preference `Leave app running in the
notification area` and set to false. Style the disabled preference
(Linux only). Style is suitable for normal mode and dark mode.
Uses an environment variable as suggested by @nathanlesage [2]

[1] https://extensions.gnome.org/extension/615/appindicator-support/
[2] https://github.com/Zettlr/Zettlr/issues/175#issuecomment-814900960

nathanlesage commented:
Hey, thanks! Looks pretty nice, especially the checking part – thank you for your work already!

One thing I'm curious about is the return value of the environment check function. I'm thinking: Wouldn't it be cleaner if the function would only return a Boolean value, and throw an error if applicable? I think this would have the following benefits:

- No custom Type (I try to limit these, since it's easy to forget where a type has been defined)
- The return value would always be a Boolean
- In case of error, we could catch that one in the actual environment check, log a warning and set said environment variable to false

Or, even better: Have the function not throw anything but immediately log anything …?

Furthermore, since you absolutely correctly identified that environment variables must be strings, I just double checked the logs, and there was an example where instead of "true" and "false" someone simply used "0" and "1". That might save some typing, since we can't do boolean checks either way with env-variable (and always need something like `=== 'false'`). But that's just a quick thought and doesn't have to be.

cskeogh commented:
Thank you for the feedback @nathanlesage

> One thing I'm curious about is the return value of the environment check function. I'm thinking: Wouldn't it be cleaner if the function would only return a Boolean value, and throw an error if applicable?

I understand. In fact our team member @ratherto had a similar comment in their review https://github.com/UofA-SPI21-team7/Zettlr/pull/22#discussion_r623168715

I think "exceptions are for exceptional circumstances" and tray supported = false is a common occurrence. Do you agree?

The reason the error is returned (rather than logged only) is this string will be the help in "show hover-over help why preference is disabled (grey-out) #14. This design allows for easy updates in the future if operating system tray requirements change.

> Furthermore, since you absolutely correctly identified that environment variables must be strings, I just double checked the logs, and there was an example where instead of "true" and "false" someone simply used "0" and "1". That might save some typing, since we can't do boolean checks either way with env-variable (and always need something like `=== 'false'`). But that's just a quick thought and doesn't have to be.

I agree. "0" and "1" is more common and more consistent. I will do this.

nathanlesage commented:
> The reason the error is returned (rather than logged only) is this string will be the help in "show hover-over help why preference is disabled (grey-out) #14. This design allows for easy updates in the future if operating system tray requirements change.

I fully understand this. However, that's what the log system is for. Whenever something goes awry, I always tell people to look into the logs, and I assume that this is what people would expect if they encounter a problem.

If you still want to use the error message in some display, that's easy:

```js
try {
  process.env.VARIABLE = await someFunctionThatThrows()
  // Throw an error like such: throw new Error('A custom error message')
  // The string will be written to err.message
} catch(err) {
  process.env.VARIABLE = '0'
  const myErrorMessage: string = err.message
  global.log.error(`[App] Could not detect libappindicator support: ${myErrorMesage}`, err)
  // Do something with the error message
}
```

cskeogh commented:
> I fully understand this. However, that's what the log system is for. Whenever something goes awry, I always tell people to look into the logs, and I assume that this is what people would expect if they encounter a problem.

Good idea. I will log the error.

Do you agree "exceptions are for exceptional circumstances" and tray supported = false is a common occurrence. Therefore 'custom type' design is ok? Or would you like an exception used?

nathanlesage commented:
> Do you agree "exceptions are for exceptional circumstances" and tray supported = false is a common occurrence. Therefore 'custom type' design is ok? Or would you like an exception used?

Yes and no. I agree that exceptions should be that – exceptional – but I don't agree that a custom type is okay in this specific case. The idea is the following: In most cases, try support is not given, but given the way we have to check that (by running a third-party binary that must be available, using a function that will throw if the binary is not available), in this case we'll have frequent `throw`s that are not exceptional in the narrow sense. Thus, in this case it's okay to catch the exceptions instead of returning a custom type. The custom type overcomplicates things where we don't have to overcomplicate them.

cskeogh commented:
> Thus, in this case it's okay to catch the exceptions instead of returning a custom type. The custom type overcomplicates things where we don't have to overcomplicate them.

Thank you for the reply. Not a problem, I'll implement an exception.

cskeogh committed:
Log warning if tray not supported

Log warning if tray not supported. Update environment variable
ZETTLR_IS_TRAY_SUPPORTED to use '1' for True and '0' for False for
consistency.

nathanlesage commented on source/app/util/environment-check.ts:
```diff
@@ -132,7 +132,14 @@ export default async function environmentCheck (): Promise<void> {
-  process.env.ZETTLR_IS_TRAY_SUPPORTED = (await isTraySupported()).value.toString()
+  try {
+    if (await isTraySupported()) {
+      process.env.ZETTLR_IS_TRAY_SUPPORTED = '1'
```
This looks much better. However, I'd still argue in here you should just put the following:

```js
process.env.ZETTLR_IS_TRAY_SUPPORTED = await isTraySupported()
```

Otherwise, when isTraySupported returns _false_, the environment variable will not even be set at all.

If you use the oneliner I provided above, however, the variable will be set, and only if the function throws an error do you have to manually set it to '0' and log the warning.

cskeogh replied:
Good idea. Your suggested code needs a slight change because can't assign `Boolean` to `String | undefined` 

``` typescript
    process.env.ZETTLR_IS_TRAY_SUPPORTED = await isTraySupported() ? '1' : '0'
```

nathanlesage replied:
Absolutely fine!

nathanlesage commented on source/app/util/environment-check.ts:
```diff
@@ -132,7 +132,14 @@ export default async function environmentCheck (): Promise<void> {
+    }
+  } catch (err) {
+    process.env.ZETTLR_IS_TRAY_SUPPORTED = '0'
+    global.log.warning(err)
```
Just a note of caution:

* global.log.warning expects a simple string
* global.log.error expects a simple string as its first argument, but supports passing in an arbitrary object (in this case: The full error object) as its second argument.

So you need to change this to one of the following:

```js
global.log.warning(err.message) // If you don't need any more information

global.log.error(err.message, err) // In case we might need the stack trace and further info from the error object
```

cskeogh replied:
Thank you, good pick-up. I chose `global.log.warning`

nathanlesage reviewed (changes requested):
We're getting there!

cskeogh committed:
Correct isTraySupported() usage and log usage

Correct isTraySupported() usage when return false. Correct
`lobal.log.warning` usage - expects a simple string.
